### DISCUSSION TITLE
Guidance on Session and Token Lifetimes

### DISCUSSION THREAD

**Discussion by ddurham2** [keycloak/keycloak#14128] [Q&A]
So I think I'm grasping what a realm's token configurable lifetimes mean.  Let's put it in terms of a web application.

We have:
* _SSO Session Max_ - The time after which a user will be absolutely automatically logged out and will require re-auth
* _SSO Session Idle_ - If the user closes the browser/powers down and leaves it off for this long, then the access tokens won't be refreshed (resetting the idle timer) and the user will be logged out due to inactivity and have to re-auth.
* _Client Session Max/Idle_ - same, but a default for individual clients, and if left unconfigured just uses the _SSO Session_ values. (right?)
 
Then we have:
* _Access Token Lifespan_ - The token used to access the web applications APIs will life only this long, and will have to be requested again (using the refresh token obtained at login, which itself expires after the _SSO Session Idle_ time).  Why have a short lifetime?  I presume because we're handing these tokens off to various places and don't want a leak of them to be as bad as a leak of the much longer lived session token.

Okay.. now I'm interested in some guidance as compared to other online services.

Take a **banking** site: Surely _SSO Session Max_ should be rather short (say 24 hours).  And the _SSO Session Idle_ time should be shorter (say 15 minutes), and the _Access Token Lifespan_ should be pretty short, say 1-5 minutes.  That makes sense, and is what I experience with online banking sites.

But take a more casual **consumer** service like Spotify, Netflix, or Pay-for AI Image generation.  The consequences of compromise are much lower.  In my experience, these sites rarely (if ever?) require to you log in again.  You conveniently open your browser and you're still in until you explicitly log out.  

I also understand that linked devices (e.g. a TV linked to your Netflix account) would be using an _Offline Access_ token, and those don't typical expire.  You can remove their access from the website usually, and those tokens have more limited rights anyway.

But in these cases, where in the web browser I seem to be logged in forever, is this just essentially a _SSO Session Max_ of several years?, and maybe a _SSO Session Idle_ of maybe 30 days or a year?   Or is it essentially an infinite _SSO Session Max_ with _SSO Session Idle_ time of several years?  What's typically the case here?

Thanks

**Comment by krishnakc1** [comment]
Some excellent explanation here -
https://stackoverflow.com/questions/52040265/how-to-specify-refresh-tokens-lifespan-in-keycloak

and details from keycloak website
https://www.keycloak.org/docs/latest/server_admin/#_timeouts
